sound/soc/sof/sof-client-probes.c

```diff
@@ -0,0 +1,796 @@
+exit:
+	kfree(tkns);
+	return ret;
+}
+
+static const struct file_operations probe_points_remove_fops = {
+	.open = simple_open,
+	.write = probe_points_remove_write,
+	.llseek = default_llseek,
+};
+
+static struct snd_soc_dai_driver sof_probes_dai_drv[] = {
+{
+	.name = "Probe Extraction CPU DAI",
+	.compress_new = snd_soc_new_compress,
+	.cops = &sof_probes_compr_ops,
+	.capture = {
+		.stream_name = "Probe Extraction",
+		.channels_min = 1,
+		.channels_max = 8,
+		.rates = SNDRV_PCM_RATE_48000,
+		.rate_min = 48000,
+		.rate_max = 48000,
+	},
+},
+};
+
+static const struct snd_soc_component_driver sof_probes_component = {
+	.name = "sof-probes-component",
+	.compress_ops = &sof_probes_compressed_ops,
+	.module_get_upon_open = 1,
+};
+
+SND_SOC_DAILINK_DEF(dummy, DAILINK_COMP_ARRAY(COMP_DUMMY()));
+
+static int sof_probes_client_probe(struct auxiliary_device *auxdev,
+				   const struct auxiliary_device_id *id)
+{
+	struct sof_client_dev *cdev = auxiliary_dev_to_sof_client_dev(auxdev);
+	struct device *dev = &auxdev->dev;
+	struct snd_soc_dai_link_component platform_component[] = {
+		{
+			.name = dev_name(dev),
+		}
+	};
+	struct snd_soc_card *card;
+	struct sof_probes_priv *priv;
+	struct snd_soc_dai_link_component *cpus;
+	struct sof_probes_pdata *pdata;
+	struct snd_soc_dai_link *links;
+	int ret;
+
+	if (!dev->platform_data) {
+		dev_err(dev, "error: missing platform data\n");
+		return -ENODEV;
+	}
+
+	priv = devm_kzalloc(dev, sizeof(*priv), GFP_KERNEL);
+	if (!priv)
+		return -ENOMEM;
+
+	pdata = dev->platform_data;
+
+	if (!pdata->ops->assign || !pdata->ops->free ||
+	    !pdata->ops->set_params || !pdata->ops->trigger ||
+	    !pdata->ops->pointer) {
+		dev_err(dev, "error: missing platform callback(s)\n");
+		return -ENODEV;
+	}
+
+	priv->pdata = pdata;
+	cdev->data = priv;
+
+	/* register probes component driver and dai */
+	ret = devm_snd_soc_register_component(dev, &sof_probes_component,
+					      sof_probes_dai_drv,
+					      ARRAY_SIZE(sof_probes_dai_drv));
+	if (ret < 0) {
+		dev_err(dev, "error: failed to register SOF probes DAI driver %d\n", ret);
+		return ret;
+	}
+
+	/* set client data */
+	priv->extractor_stream_tag = SOF_PROBES_INVALID_NODE_ID;
+
+	/* create probes debugfs dir under SOF debugfs root dir */
+	priv->dfs_root = debugfs_create_dir("probes",
+					    sof_client_get_debugfs_root(cdev));
+	if (!IS_ERR_OR_NULL(priv->dfs_root)) {
+		/* create read-write probes_points debugfs entry */
+		debugfs_create_file("probe_points", 0644, priv->dfs_root, cdev,
+				    &probe_points_fops);
+
+		/* create read-write probe_points_remove debugfs entry */
+		debugfs_create_file("probe_points_remove", 0644, priv->dfs_root,
+				    cdev, &probe_points_remove_fops);
+	}
+
+	links = devm_kcalloc(dev, SOF_PROBES_NUM_DAI_LINKS, sizeof(*links), GFP_KERNEL);
+	cpus = devm_kcalloc(dev, SOF_PROBES_NUM_DAI_LINKS, sizeof(*cpus), GFP_KERNEL);
+	if (!links || !cpus) {
+		debugfs_remove(priv->dfs_root);
```

Comment:
previously there used to be a `priv->dfs_root = NULL;` line below this one. You probably removed it, because on this path probing will fail and all resources, allocated for `auxdev->dev` will be freed immediately anyway

Reply:
Let me check this, thanks!

Reply:
Yes, there is no need to NULL it as we return right away with -ENOMEM from the probe function.